### PR TITLE
Do not create any instance manager pods for v2 volumes if one is already existing

### PR DESCRIPTION
**Pull request by derekbit**
#### Which issue(s) this PR fixes:

Longhorn/longhorn#8456

#### What this PR does / why we need it:

An extra instance manager pod for v2 data engine requires an additional CPU core and 2 GiB of memory. If the resources are insufficient, the instance manager pod will be recreated and retried. To minimize unnecessary retries, do not create any new instance manager pods if one is already running.

### CODE REVIEW

**Comment by derekbit** [comment]
Ping @shuo-wu @mantissahz @c3y1huang @innobead for review. Thank you.

**Comment by derekbit** [comment]
Updated and ready for review.
cc  @shuo-wu @c3y1huang @innobead

**Comment by derekbit** [comment]
@c3y1huang Can you help review the PR? Thank you.

File: controller/instance_manager_controller.go [hunk 1]
```diff
@@ -334,6 +334,21 @@ func (imc *InstanceManagerController) syncInstanceManager(key string) (err error
 		}
 	}()
 
+	// An instance manager pod for v2 volume need to consume huge pages, and disks managed by the
+	// pod is unable to managed by another pod. Therefore, if an instance manager pod is running on a node,
+	// an extra instance manager pod for v2 volume should not be created.
+	if types.IsDataEngineV2(im.Spec.DataEngine) {
+		if im.Status.CurrentState == longhorn.InstanceManagerStateStopped {
```
Comment (c3y1huang) [inline controller/instance_manager_controller.go:341]: Is there a reason the state is being checked both here and again in `canProceedWithInstanceManagerSync`?

Comment (derekbit) [reply controller/instance_manager_controller.go]: Refactored the codes. Can you help review it again? Thank you.

File: controller/instance_manager_controller.go [hunk 2]
```diff
@@ -365,6 +378,36 @@ func (imc *InstanceManagerController) syncInstanceManager(key string) (err error
...
+	// If the instance manager is not stopped, proceed with the sync.
+	if currentIm.Status.CurrentState != longhorn.InstanceManagerStateStopped {
+		return true, nil
+	}
+
+	ims, err := imc.ds.ListInstanceManagersByNodeRO(currentIm.Spec.NodeID, longhorn.InstanceManagerTypeAllInOne, longhorn.DataEngineTypeV2)
+	if err != nil {
+		return false, err
+	}
+	// If there is another non-stooped instance manager pod for v2 volume, do not proceed with the sync.
```
[the remark this answers is not in the document]
Comment (derekbit) [reply controller/instance_manager_controller.go]: Fixed. Thank you.

**Review by c3y1huang — APPROVED** [review]

**Pull request merged**